dali/operators/reader/loader/video/frames_decoder.cc

```diff
@@ -236,20 +243,97 @@ FramesDecoder::FramesDecoder(const char *memory_file, int memory_file_size, bool
+      nullptr,
+      detail::seek_memory_video_file);
+
+    av_state->ctx_->pb = av_io_context;
+
+    int ret = avformat_open_input(&av_state->ctx_, "", nullptr, nullptr);
+    DALI_ENFORCE(
+      ret == 0,
+      make_string(
+        "Failed to open video file ",
+        Filename(),
+        "due to ",
+        detail::av_error_string(ret)));
+    av_state->stream_id_ = av_find_best_stream(
+      av_state->ctx_, AVMEDIA_TYPE_VIDEO, -1, -1, nullptr, 0);
+    av_state->codec_params_ = av_state->ctx_->streams[av_state->stream_id_]->codecpar;
+
+    av_state->codec_ctx_ = avcodec_alloc_context3(av_state->codec_);
+    DALI_ENFORCE(av_state->codec_ctx_, "Could not alloc av codec context");
+
+    ret = avcodec_parameters_to_context(av_state->codec_ctx_, av_state->codec_params_);
+    DALI_ENFORCE(
+      ret >= 0,
+      make_string("Could not fill the codec based on parameters: ", detail::av_error_string(ret)));
+
+    av_state->packet_ = av_packet_alloc();
+    DALI_ENFORCE(av_state->packet_, "Could not allocate av packet");
+}
+
 void FramesDecoder::ParseNumFrames() {
   int curr_num_frames = 0;
-  while (av_read_frame(av_state_->ctx_, av_state_->packet_) >= 0) {
-    // We want to make sure that we call av_packet_unref in every iteration
-    auto packet = AVPacketScope(av_state_->packet_, av_packet_unref);
 
-    if (packet->stream_index != av_state_->stream_id_) {
-      continue;
+  if (IsFormatSeekable()) {
+    while (av_read_frame(av_state_->ctx_, av_state_->packet_) >= 0) {
+      // We want to make sure that we call av_packet_unref in every iteration
+      auto packet = AVPacketScope(av_state_->packet_, av_packet_unref);
+
+      if (packet->stream_index != av_state_->stream_id_) {
+        continue;
+      }
+      curr_num_frames++;
+    }
+
+    num_frames_ = curr_num_frames;
+    Reset();
+  } else {
+    // Failover for unseekable video
+    auto current_position = memory_video_file_->position_;
+    memory_video_file_->Seek(0, SEEK_SET);
+    std::unique_ptr<AvState> tmp_av_state = std::make_unique<AvState>();
+    CreateAvState(tmp_av_state, false);
+
+    while (av_read_frame(tmp_av_state->ctx_, tmp_av_state->packet_) >= 0) {
+      // We want to make sure that we call av_packet_unref in every iteration
+      auto packet = AVPacketScope(tmp_av_state->packet_, av_packet_unref);
+
+      if (packet->stream_index != tmp_av_state->stream_id_) {
+        continue;
+      }
+      curr_num_frames++;
+    }
+
+    num_frames_ = curr_num_frames;
+    memory_video_file_->Seek(current_position, SEEK_SET);
+
+    if (tmp_av_state->packet_->pts == AV_NOPTS_VALUE) {
+      // zero_latency_ = false;
     }
-    curr_num_frames++;
   }
+}
 
-  num_frames_ = curr_num_frames;
-  Reset();
+bool FramesDecoder::IsFormatSeekable() {
+  if (
+    av_state_->ctx_->iformat->read_seek == nullptr &&
+    av_state_->ctx_->iformat->read_seek2 == nullptr) {
+    return false;
+  }
+
+  return av_state_->ctx_->pb->read_seek != nullptr;
```

Comment:
So still if one of `iformat->read_seek` and `iformat->read_seek2` is not null we still need to check `pb->read_seek`?

Reply:
Yes, we need both (one of the read functions to be not null and `pb->read_seek` to be not null) to be able to seek. I was not able to 100% confirm that in all possible scenarios one of these implies the other, so I decided to just check for both.